overlord/state/taskrunner.go

```diff
@@ -73,13 +85,15 @@ func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 
 		r.state.Lock()
 		defer r.state.Unlock()
+		halted := task.HaltTasks()
 		if err == nil {
 			task.SetStatus(DoneStatus)
 		} else {
-			task.SetStatus(ErrorStatus)
+			taskFail(task)
+		}
+		if len(halted) > 0 {
+			r.state.EnsureBefore(0)
```

Comment:
ah, what we could is do this only for the done case, but the next branch has more things to do in Ensure also when we marked things in error